fetch/fetch.go

```diff
@@ -0,0 +1,556 @@
+// Package fetch contains mechanism to fetch data from remote peersProvider
+package fetch
+
+import (
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/log"
+	p2pconf "github.com/spacemeshos/go-spacemesh/p2p/config"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	p2ppeers "github.com/spacemeshos/go-spacemesh/p2p/peers"
+	"github.com/spacemeshos/go-spacemesh/p2p/server"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"github.com/spacemeshos/go-spacemesh/rand"
+	"sync"
+	"time"
+)
+
+type (
+	// DataCallbackHandler is the callback that will be called when `hash` data is received, hash of the bytes must match
+	// hash
+	DataCallbackHandler func(hash types.Hash32, buf []byte) error
+)
+
+// Hint marks which DB should be queried for a certain provided hash
+type Hint string
+
+// priority defines whether data will be fetched at once or batched and waited for up to "batchTimeout"
+// until fetched
+type priority uint16
+
+// Message priority constants
+const (
+	// Low will perform batched calls
+	Low = 0
+	// High will call fetch immediately
+	High = 1
+)
+
+const (
+	fetch         server.MessageType = 1
+	fetchProtocol                    = "/sync/2.0/"
+
+	batchMaxSize = 20
+)
+
+// ErrCouldNotSend is a special type of error indicating fetch could not be done because message could not be sent to peersProvider
+type ErrCouldNotSend error
+
+// Fetcher is the general interface of the fetching unit, capable of requesting bytes that corresponds to a hash
+// from other remote peersProvider.
+type Fetcher interface {
+	GetHash(hash types.Hash32, h Hint, dataCallback DataCallbackHandler, validateAndSubmit bool) chan HashDataPromiseResult
+	GetAllHashes(hash []types.Hash32, hint Hint, hashValidationFunction DataCallbackHandler, validateAndSubmit bool) error
+}
+
+/// request contains all relevant data for a single request for a specified hash
+type request struct {
+	successCallback      DataCallbackHandler        // successCallback is the callback called if data was retrieved successfully
```

Comment:
Naming it `onSuccess` might be clearer (also to imply that this is not about validating the success).